src/messaging/ExchangeContext.cpp

```diff
@@ -531,57 +531,60 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
         MessageHandled();
     });
 
-    if (mDispatch.IsReliableTransmissionAllowed() && !IsGroupExchangeContext())
+    if (GetSessionHandle()->RequireMRP())
     {
-        if (!msgFlags.Has(MessageFlagValues::kDuplicateMessage) && payloadHeader.IsAckMsg() &&
-            payloadHeader.GetAckMessageCounter().HasValue())
+        if (mDispatch.IsReliableTransmissionAllowed() && !IsGroupExchangeContext())
         {
-            HandleRcvdAck(payloadHeader.GetAckMessageCounter().Value());
+            if (!msgFlags.Has(MessageFlagValues::kDuplicateMessage) && payloadHeader.IsAckMsg() &&
+                payloadHeader.GetAckMessageCounter().HasValue())
+            {
+                HandleRcvdAck(payloadHeader.GetAckMessageCounter().Value());
+            }
+
+            if (payloadHeader.NeedsAck())
+            {
+                // An acknowledgment needs to be sent back to the peer for this message on this exchange,
+                HandleNeedsAck(messageCounter, msgFlags);
+            }
         }
 
-        if (payloadHeader.NeedsAck())
+        if (IsAckPending() && !mDelegate)
         {
-            // An acknowledgment needs to be sent back to the peer for this message on this exchange,
-            HandleNeedsAck(messageCounter, msgFlags);
+            // The incoming message wants an ack, but we have no delegate, so
+            // there's not going to be a response to piggyback on.  Just flush the
+            // ack out right now.
+            ReturnErrorOnFailure(FlushAcks());
         }
-    }
 
-    if (IsAckPending() && !mDelegate)
-    {
-        // The incoming message wants an ack, but we have no delegate, so
-        // there's not going to be a response to piggyback on.  Just flush the
-        // ack out right now.
-        ReturnErrorOnFailure(FlushAcks());
-    }
-
-    // The SecureChannel::StandaloneAck message type is only used for MRP; do not pass such messages to the application layer.
-    if (isStandaloneAck)
-    {
-        return CHIP_NO_ERROR;
-    }
+        // The SecureChannel::StandaloneAck message type is only used for MRP; do not pass such messages to the application layer.
+        if (isStandaloneAck)
+        {
+            return CHIP_NO_ERROR;
+        }
 
-    // Since the message is duplicate, let's not forward it up the stack
-    if (isDuplicate)
-    {
-        return CHIP_NO_ERROR;
-    }
+        // Since the message is duplicate, let's not forward it up the stack
+        if (isDuplicate)
```

Comment:
This is wrong.  The duplicate check needs to be independent of MRP.

In general, it's worth thinking a bit about which of these parts should be conditioned on what.  For example, `IsAckPending()` is going to be false if the transport does not support MRP, so there's no reason to put it inside the check...  On the other hand, the question of what to do if the other side asks for an ack is an interesting one.  Right now I guess we will ack things over BTP if the other side asks us to...

Similarly, given the ExchangeManager changes, IsEphemeralExchange() will always be false in the TCP case, right?

Reply:
Yes, you are right. Thanks. I was careful enough to keep duplicate handling out of this at the ExchangeMgr level but overlooked this one in ExchangeContext. Will take that check out.
On the other side asking for an Ack, I was thinking that it should not do so given how we structure the code on both sender and receiver side. There is no point in that over TCP and allowing that would necessarily mean allowing all of MRP over TCP which is a big waste of resources, right? 
IsEphemeralExchange() should also be false over TCP. I don't see any use for it over TCP. But it can stay inside since we do not want to check it for all TCP messages.

Reply:
> I was thinking that it should not do so

If the other side is the Matter SDK, absolutely.

If the other side is just sending various things, we just need to handle it somewhat gracefully.  How gracefully is up to us.

> IsEphemeralExchange() should also be false over TCP

Right, so we don't have to put it under the "is tcp" bit.

My main point is this: TCP is not special.  Whatever is happening for TCP should happen for BTP too.  Our existing code should generally all work with this, and if it does not, we should fix those bugs... But the fixes should not be TCP-specific.

Reply:
Agreed that TCP and BTP should follow similar flows and both should essentially exclude MRP retransmission logic. Not sure about the other side sending various things, but the receiving stack should ignore and bypass all MRP logic when handling messages over connection-oriented transports. If the other side is not Matter, and it is requesting an MRP Ack over TCP(by whatever means), the local side need not oblige and send back any acknowledgments, right?

>Right, so we don't have to put it under the "is tcp" bit.

Do you mean under the AllowsMRP() check? I have put it outside currently. But I was of the opinion that all these checks are relevant to only MRP traffic and TCP|BTP messages should not need to go through these checks. From that standpoint, only IsDuplicate() should be outside, isn't it?

Reply:
Agreed that TCP and BTP should follow similar flows and both should essentially exclude MRP retransmission logic. Not sure about the other side sending various things, but the receiving stack should ignore and bypass all MRP logic when handling messages over connection-oriented transports. If the other side is not Matter, and it is requesting an MRP Ack over TCP(by whatever means), the local side need not oblige and send back any acknowledgments, right?

>Right, so we don't have to put it under the "is tcp" bit.

Do you mean under the AllowsMRP() check? I have put it outside currently. But I was of the opinion that all these checks are relevant to only MRP traffic and TCP|BTP messages should not need to go through these checks. From that standpoint, only IsDuplicate() should be outside, isn't it?

Reply:
> Do you mean under the AllowsMRP() check? 

Right.

I don't feel too strongly about the ephemeral exchange check; if the code is simpler with it inside the `AllowsMRP()` block, great. @pidarped